packages/rest/src/http-handler.ts

```diff
@@ -33,8 +34,12 @@ export class HttpHandler {
     this.handleRequest = (req, res) => this._handleRequest(req, res);
   }
 
-  registerController(name: ControllerClass, spec: ControllerSpec) {
-    this._routes.registerController(name, spec);
+  registerController<T>(
+    controllerCtor: ControllerClass<T>,
+    spec: ControllerSpec,
+    controllerFactory?: ControllerFactory<T>,
```

Comment:
I find it a bit weird to have two controller-related arguments interleaved with other non-controller arguments.

I am proposing to reorder the arguments of this method as follows:

```ts
registerController<T>(
    spec: ControllerSpec,
    controllerCtor: ControllerClass<T>,
    controllerFactory?: ControllerFactory<T>,
) {}
```

The same comment applies to other places where `controllerFactory` argument as added to a function/method accepting `controllerCtor`.

I would also like to make `controllerFactory` as a required argument, and treat `controllerCtor` more like a reference for obtaining controller metadata. This will open more possibilities for us, for example allow `controllerFactory` to return a plain object (no classes involved) and then craft `controllerCtor` in such way that allows the rest of the framework to fetch the required metadata.

If we agree on this second change, then I would swap the order of factory/ctor argument:

```ts
registerController<T>(
    spec: ControllerSpec,
    controllerFactory?: ControllerFactory<T>,
    controllerCtor: ControllerClass<T>,
) {}
```

This will also allow us to preserve required `methodName` while grouping controller-related args in `route<T>()` API below.

```ts
route<T>(
   verb: string,
   path: string,
   spec: OperationObject,
   controllerFactory: ControllerFactory<T>,
   controllerCtor: ControllerClass<T>,
   methodName: string,
){}
```

Thoughts?

Reply:
I move `controllerFactory` to be next to `controllerCtor`.

In most cases, we have `ControllerClass` and the controller factory is for the class. With the intention to customize by exception, I leave controllerFactory to be after controlCtor and optional in most cases.